File: Software tests/IIR.py
class IIR:
    def __init__(self,a:list,b:list):
        self.a = a
        self.b = b
        self.x = [0 for _ in range(len(b))]
        self.y = [0 for _ in range(len(a))]
        self.out = 0
    
    def update(self,input:float) -> float:
        
        # Shift input and output arrays
        self.x.insert(0,input)
        self.x.pop(len(self.x)-1)

        self.y.insert(0,None)
        self.y.pop(len(self.y)-1)

        self.out = 0
        for i in range(len(self.b)):
            self.out += self.b[i]*self.x[i]
        for i in range(1,len(self.a)):
            self.out -= self.a[i]*self.y[i]

        self.y[0] = self.out

        return float(self.out)

if __name__ == "__main__":

    from matplotlib import pyplot as plt

    # w0 = 2.35 hz, zeta = 0.01, ts = 0.01
    b = [1, -1.9780,0.9966]
    a = [1, -1.7276,0.7462]

    notch = IIR(a,b)

    impulse = [1]
    impulse += [0 for _ in range(99)]

    step = [1 for _ in range(100)]

    out = []
    for i in impulse:
        out.append(notch.update(i))

    plt.plot([x/len(out) for x in range(len(out))],out)
    plt.grid()
    plt.show()

File: Software tests/iir_c_test/iir.cpp
#include"iir.h"

/*
Basic 3-parameter IIR-filter of the format:

b[0] + b[1]*z^-1 + b[2]*z^-2 
----------------------------
a[0] + a[1]*z^-1 + a[2]*z^-2

*/
IIR::IIR(float a_in[3], float b_in[3]){

    a[0] = a_in[0]; a[1] = a_in[1]; a[2] = a_in[2];
    b[0] = b_in[0]; b[1] = b_in[1]; b[2] = b_in[2];

    x[0] = 0; x[1] = 0; x[2] = 0; 
    y[0] = 0; y[1] = 0; y[2] = 0;

}

float IIR::update(float input){
    x[2] = x[1]; x[1] = x[0];
    x[0] = input;

    y[2] = y[1]; y[1] = y[0];
    
    float output = 
        + x[0]*b[0]
        + x[1]*b[1]
        + x[2]*b[2]

        - y[1]*a[1]
        - y[2]*a[2] ;

    output = output/a[0];

    y[0] = output;

    return output;
}